posthog/temporal/llm_analytics/trace_clustering/activities.py

```diff
@@ -0,0 +1,170 @@
+from posthog.temporal.llm_analytics.trace_clustering.clustering import (
+    calculate_trace_distances,
+    perform_kmeans_with_optimal_k,
+    select_representatives_from_distances,
+)
+from posthog.temporal.llm_analytics.trace_clustering.data import fetch_trace_embeddings_for_clustering
+from posthog.temporal.llm_analytics.trace_clustering.event_emission import emit_cluster_events
+from posthog.temporal.llm_analytics.trace_clustering.labeling import generate_cluster_labels
+from posthog.temporal.llm_analytics.trace_clustering.models import (
+    ClusteringActivityInputs,
+    ClusteringComputeResult,
+    ClusteringMetrics,
+    ClusteringResult,
+    EmitEventsActivityInputs,
+    GenerateLabelsActivityInputs,
+    GenerateLabelsActivityOutputs,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def _perform_clustering_compute(inputs: ClusteringActivityInputs) -> ClusteringComputeResult:
+    """CPU-bound compute: fetch embeddings, cluster, compute distances.
+
+    This is the synchronous implementation called by the activity.
+    """
+    window_start = parse_datetime(inputs.window_start)
+    window_end = parse_datetime(inputs.window_end)
+    if window_start is None or window_end is None:
+        raise ValueError(f"Invalid datetime format: window_start={inputs.window_start}, window_end={inputs.window_end}")
+
+    clustering_run_id = f"team_{inputs.team_id}_{window_end.isoformat()}"
+
+    # Fetch trace IDs and embeddings
+    trace_ids, embeddings_map = fetch_trace_embeddings_for_clustering(
+        team_id=inputs.team_id,
+        window_start=window_start,
+        window_end=window_end,
+        max_samples=inputs.max_samples,
+    )
+
+    embeddings_array = np.array(list(embeddings_map.values()))
+
+    # Perform clustering with optimal k selection
+    kmeans_result = perform_kmeans_with_optimal_k(embeddings_array, inputs.min_k, inputs.max_k)
+
+    # Compute distance matrix
+    distances_matrix = calculate_trace_distances(embeddings_array, np.array(kmeans_result.centroids))
+
+    # Select representative traces for LLM labeling
+    representative_trace_ids = select_representatives_from_distances(
+        labels=np.array(kmeans_result.labels),
+        distances_matrix=distances_matrix,
+        trace_ids=trace_ids,
+        n_closest=constants.DEFAULT_TRACES_PER_CLUSTER_FOR_LABELING,
+    )
+
+    return ClusteringComputeResult(
+        clustering_run_id=clustering_run_id,
+        trace_ids=trace_ids,
+        labels=kmeans_result.labels,
+        centroids=kmeans_result.centroids,
+        distances=distances_matrix.tolist(),
+        representative_trace_ids=representative_trace_ids,
+    )
+
+
+@activity.defn
+async def perform_clustering_compute_activity(inputs: ClusteringActivityInputs) -> ClusteringComputeResult:
+    """Activity 1: CPU-bound compute - fetch embeddings, cluster, compute distances.
+
+    This activity handles all the compute-intensive work:
```

Comment:
Is the reason why you kept the fetching of embeddings as part of this activity because you don't want to pass the 3-4 MB between activities?

I would have thought that:

Activity 1:
- Fetches embeddings from ClickHouse

Activity 2:
- Performs k-means clustering with optimal k selection
- Calculates distances from each trace to all centroids
- Selects representative traces for labeling

would have made more sense in terms of separating things up.

If something is wrong in the clustering calculation, we wouldn't have to query the DB again for the embeddings.

Is passing that much data between activities a bad practice for Temporal?